### ISSUE TITLE
__getitem__  and __setitem__ assume DNDarray is balanced

### ISSUE THREAD

**Issue by ClaudiaComito** [helmholtz-analytics/heat#668]
**Description**
`dndarray.__getitem__` and `dndarray.__setitem__` rely on  `comm.chunk()` to provide the information about what slice of the data is on each process. However `comm.chunk()` only calculates such slice, on the assumption that the underlying array is balanced. 

A very common use case for working with unbalanced dndarrays is `ht.where()`. The result will almost always be unevenly distributed, and we want to keep it that way because it matches the relevant data on the node.

Here's an example of the problem (`mpirun -n 2`):
```
ht_mask_matrix = ht.random.randint(0,2,(13,10), split=0)
ht_where = ht.where(ht_mask_matrix > 0)

>> [1,1]<stdout>:ht_where.gshape, ht_where.lshape =  (70, 2) (32, 2)
>> [1,0]<stdout>:ht_where.gshape, ht_where.lshape =  (70, 2) (38, 2) <---

ht_where_x, ht_where_y = ht_where[:, 0], ht_where[:, 1]

>> [1,1]<stdout>:ht_where_x.lshape =  (32,)
>> [1,0]<stdout>:ht_where_x.lshape =  (35,) <---
```

**To Reproduce**
see above

**Expected behavior**
The getters and setters should know know the actual slice of data are on each node. The least disruptive solution for me would be to have `comm.chunk()` check for balance, and collect the actual chunking information (instead of calculating it) if the array is unbalanced.

**Version Info**
0.4. current main branch

**Comment by ClaudiaComito** [comment]
@coquelin77 on MM points to the communication overhead that would arise by checking for balance every time we run comm.chunk().

I'm thinking: how about a `dndarray.balanced` boolean attribute?  
- it would be set, mostly to `True`, by factories and operations (notable exceptions: `ht.where`, `ht.flatten` maybe? @lenablind )
- it would save us a ton of `is_balanced()` communication
- `comm.chunk()` would work as usual `if self.balanced`, i.e. in the vast majority of cases, otherwise collect and return information on local data slices
- getters and setters could return the correct slice no matter what

Thoughts?

**Comment by lenablind** [comment]
> @coquelin77 on MM points to the communication overhead that would arise by checking for balance every time we run comm.chunk().
> 
> I'm thinking: how about a `dndarray.balanced` boolean attribute?
> 
>     * it would be set, mostly to `True`, by factories and operations (notable exceptions: `ht.where`, `ht.flatten` maybe? @lenablind )
> 
>     * it would save us a ton of `is_balanced()` communication
> 
>     * `comm.chunk()` would work as usual `if self.balanced`, i.e. in the vast majority of cases, otherwise collect and return information on local data slices
> 
>     * getters and setters could return the correct slice no matter what
> 
> 
> Thoughts?

I think it seems like a good idea. Especially the communication point is quite convincing as it might save time. Though I don't think we need to take care about the exceptions (as the attribute could be set to True by calling `balance_()`). `ht.flatten` wouldn't be affected either way as the result is balanced within HeAT. Nevertheless, we would need to make sure that the attribute can't be accessed/set outside of the `balance` function itself as that would probably have quite an impact. The only thing that leaves me critical is that this attribute won't likely be used in most of the cases (apart from `._balance` for sure). But I guess the overhead regarding a boolean wouldn't be that large.

**Comment by ClaudiaComito** [comment]
Fixed in #758
